src/Bicep.Core/Semantics/Namespaces/INamespaceProviderManager.cs

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+
+namespace Bicep.Core.Semantics.Namespaces;
+
+public interface INamespaceProviderManager
```

Comment:
`INamespaceProvider` is already acting a bit like a factory class for `NamespaceType` - do we need this extra level of indirection (factoryfactory vs factory)?

Reply:
Potentially, though the methods of `NamespaceProvider` would then need to accept a template URI as an argument. This PR includes making a factory for IFeatureProvider, which needs to be in place for #7955, and the behavior of `NamespaceProvider` varies based on which features are enabled. Using the factoryfactory pattern ( :( ) lets the variation be resolved once per file in a given compilation (as opposed to each time a `NamespaceProvider` method is invoked).

That said, factoryfactories strike me as a design smell, and I'm introducing two new ones in this PR (`NamepaceProviderManager` and `ApiVersionProviderManager`).

Reply:
I found a way to remove `NamespaceProviderManager` and `ApiVersionProviderManager`, which also made this PR significantly smaller!